lib/src/tools/format_tool.dart

```diff
@@ -161,20 +198,49 @@ class FormatTool extends DevTool {
       if (hiddenIndex != null) {
         final hiddenDirectory = p.joinAll(parts.take(hiddenIndex + 1));
         hiddenDirectories.add(hiddenDirectory);
+        _log.finest('skipping hidden dir $hiddenDirectory');
+        if (collapseDirectories) {
+          currentDirectory = relative;
+          skipFilesInDirectory = true;
+        }
         continue;
       }
 
       if (exclude.any((glob) => glob.matches(relative))) {
+        _log.finest('excluding $relative');
         excludedFiles.add(relative);
       } else {
-        if (entry is File) includedFiles.add(relative);
+        if (collapseDirectories && entry is Directory) {
+          _log.finest('directory: $entry');
+          currentDirectory = relative;
+          // It seems we can rely on the order of files coming from listSync.
+          // If a directory does not match any of the globs, and does not contain any of the globs,
+          // we should be able to just add that directory and skip adding any of its children files or directories.
+          if (exclude.any((glob) => p.isWithin(
+              entry.path, p.relative(glob.toString(), from: dir.path)))) {
```

Comment:
I don't think this works. For one, it's treating `glob.toString()` as a filepath, when it's actually a pattern. For example, it's common to exclude generated files with a `Glob('**.g.dart')` glob which matches any file in the project with a `.g.dart` extension. But used here as a filepath, you'd get false negatives:

```dart
p.isWithin('lib/', '**.g.dart') // false
```

Another example would be using `{}` to list multiple options:

```dart
p.isWithin('lib/src/', 'lib/{src,internal}/*.dart') // false
```

In both of these cases, we should be matching but aren't.

---

To fully support this functionality, I think we'd need to expand the list of files and then go back and test them against the exclude globs using `Glob.matches()`, collapsing directories one at a time if none of its children match.

Reply:
This implementation happens to work in DPC because the excludes are only using patterns to target whole directories (e.g. `app/build/**`), and the rest are actual file paths.